src/services/node-checker-wrapper.ts

```diff
@@ -0,0 +1,272 @@
+  constructor(pocket: Pocket, redis: Redis, metricsRecorder: MetricsRecorder, origin: string) {
+    this.pocket = pocket
+    this.redis = redis
+    this.metricsRecorder = metricsRecorder
+    this.origin = origin
+  }
+
+  /**
+   * Helper method to check for cached checks, is a check is already cached or in progress,
+   * returns valid nodes from the cached result or session. Otherwise set a cache lock.
+   * @param {Node[]} nodes session nodes.
+   * @param cacheKey key to get/set results.
+   * @returns nodes cached or provided on case of cache/lock.
+   */
+  protected async cacheNodes(nodes: Node[], cacheKey: string): Promise<Node[]> {
+    const checkedNodes: Node[] = []
+    let checkedNodesList: string[] = []
+
+    const checkedNodesCached = await this.redis.get(cacheKey)
+
+    checkedNodesList = JSON.parse(checkedNodesCached)
+
+    if (checkedNodesList && checkedNodesList.length > 0) {
+      for (const node of nodes) {
+        if (checkedNodesList.includes(node.publicKey)) {
+          checkedNodes.push(node)
+        }
+      }
+      return checkedNodes
+    }
+
+    // Cache is stale, start a new cache fill
+    // First check cache lock key; if lock key exists, return full node set
+    const chainLock = await this.redis.get('lock-' + cacheKey)
+
+    if (chainLock) {
+      return nodes
+    } else {
+      // Set lock as this thread checks the Chain with 60 second ttl.
+      // If any major errors happen below, it will retry the check every 60 seconds.
+      await this.redis.set('lock-' + cacheKey, 'true', 'EX', 60)
+    }
+
+    return checkedNodes
+  }
+
+  /**
+   * Logs responses from the node checks and filters failing nodes.
+   * @param checkType type of check made.
+   * @param nodes nodes to be filtered.
+   * @param nodesPromise results of a check on node's.
+   * @param blockchainID blockchain used for the checks.
+   * @param requestID request id.
+   * @param relayStart time when the checks started.
+   * @param applicationID application database's ID.
+   * @param applicationPublicKey application's public key.
+   * @returns nodes having successful and valid check results.
+   */
+  protected async filterNodes<T>(
+    checkType: Check,
+    nodes: Node[],
+    nodesPromise: PromiseSettledResult<NodeCheckResponse<T>>[],
+    blockchainID: string,
+    pocketSession: Session,
+    requestID: string,
+    relayStart: [number, number],
+    applicationID: string,
+    applicationPublicKey: string
+  ): Promise<NodeCheckResponse<T>[]> {
+    const filteredNodes: NodeCheckResponse<T>[] = []
+    const { sessionNodes } = pocketSession
+    const sessionHash = hashBlockchainNodes(blockchainID, sessionNodes)
+
+    for (const [idx, nodeCheckPromise] of nodesPromise.entries()) {
+      const node = nodes[idx]
+
+      const { serviceURL, serviceDomain } = await getNodeNetworkData(this.redis, node.publicKey, requestID)
+
+      // helps debugging
+      const formattedType = checkType.replace('-', ' ').toUpperCase()
```

Comment:
Wonder if instead of doing this we should just have an object (or a set of `Symbol`s) with the different check types? That way it's also impossible to fat-finger them.

Reply:
they're part of a type [https://github.com/pokt-foundation/portal-api/blob/54afc9fd085f04337e7f5fdf4ce9c2b78ad35c1f/src/services/node-checker.ts#L6](from node-checker) writing improperly will trigger a compiler error and autocomplete on an editor with intellisense-like features

Reply:
Yup, if it's a type sg as well